### PR TITLE
Review Leader QoS

### PR DESCRIPTION
**Pull request by taozhu-chicago**
#### Problem
review Leader QoS Proposal

#### Summary of Changes

related to #23211 

Fixes #

### CODE REVIEW

**Comment by taozhu-chicago** [comment]
commit #a277cd871ab730d72b2e0cc420bfd5322d0ef939

File: docs/src/proposals/leader-qos.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+   Where forwarding is a process of select unprocessed packets from unforwarded packet_batches in FIFO order, batch-send to Fwd ports. 
+
+The aim of this proposal is to replace linear FIFO buffer traversing with transaction-based prioritization. The change scope is mainly at step #3, instead of linearly selecting unprocessed packets from buffered packet_batches, it will sort unprocessed packets with following criteria to prioritize consuming or forwarding. 
+
+The consuming and forwarding processing as described above largely remains the same, as they are outside scope of this proposal. 
+
+## Prioritization criteria:
+1. Transaction's fee-per-CU ( `= (additional_fee + base_fee)/requested_CU)` ), higher paid transactions have priority to be first considered for block inclusion or forwarding. 
+2. When fee-per-CU is same, Transactions from higher staked validators have higher chance to be considered for blocking inclusion or forwarding.
```
Comment (buffalu) [inline docs/src/proposals/leader-qos.md:22]: was this in your original pr? thought it was stake weight first but maybe I misread.
this makes sense to me as opposed to pure stake weighted priority

File: docs/src/proposals/leader-qos.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+   Where consuming is a process of preparing selected packets into batches of sanitized_transactions, submitting for attempted bank execution, which in turn includes block cost tracking, account locking, and executing and committing. Retryable transactions will be put back to buffer for next iteration.
+   Where forwarding is a process of select unprocessed packets from unforwarded packet_batches in FIFO order, batch-send to Fwd ports. 
+
+The aim of this proposal is to replace linear FIFO buffer traversing with transaction-based prioritization. The change scope is mainly at step #3, instead of linearly selecting unprocessed packets from buffered packet_batches, it will sort unprocessed packets with following criteria to prioritize consuming or forwarding. 
+
+The consuming and forwarding processing as described above largely remains the same, as they are outside scope of this proposal. 
+
+## Prioritization criteria:
+1. Transaction's fee-per-CU ( `= (additional_fee + base_fee)/requested_CU)` ), higher paid transactions have priority to be first considered for block inclusion or forwarding.
```
Comment (buffalu) [inline docs/src/proposals/leader-qos.md:21]: which cu are we talking about here?

File: docs/src/proposals/leader-qos.md (hunk 1 above)
Comment (taozhu-chicago) [reply docs/src/proposals/leader-qos.md:22]: It's part of original PR, probably I haven't fully committed the changes when you looked at it. 
Yeah, the prioritizing by fee/CU is better.

File: docs/src/proposals/leader-qos.md (hunk 2 above)
Comment (taozhu-chicago) [reply docs/src/proposals/leader-qos.md:21]: Right now, it is using `compute_budget` request CU @ https://github.com/solana-labs/solana/blob/master/program-runtime/src/compute_budget.rs#L117-L119

**Review by t-nelson — COMMENTED** [review]
Can you wrap lines at 80 char to ease commenting?

**Comment by taozhu-chicago** [comment]
> Can you wrap lines at 80 char to ease commenting?

done `:set colorcolumn=80`

File: docs/src/proposals/leader-qos.md [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+   sanitized_transactions, submitting for attempted bank execution, which in 
+   turn includes block cost tracking, account locking, and executing/committing.
+   Retryable transactions will be put back to buffer for next iteration.
+
+   "forwarding" is a process of preparing selected unprocessed packets from
+   unforwarded packet_batches, batch-send to Fwd ports. 
+
+The aim of this proposal is to enhance the packet selection process for
+"consuming" and "forwarding" within each banking_stage threads; by 
+compartmentalizing hcanges to speed up delivery. A more holistic change
```
Comment (aeyakovenko) [inline docs/src/proposals/leader-qos.md:26]: s/hcanges/changes

**Review by aeyakovenko — APPROVED** [review]

File: docs/src/proposals/leader-qos.md [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+2. When fee-per-CU is same, Transactions from higher staked validators have
+   higher chance to be considered for blocking inclusion or forwarding.
+
+## Implementation 
+Because received packets are buffered as packet_batches, in order to 
+prioritize transaction-by-transaction without additional copying of packets,
+there are few extra required changes:
+1. Deserialize packets into transactions upfront, before buffer traversing;
+2. Introduce a way (eg., locator) of iterating packets from buffered 
+   packet_batches;
```
Comment (buffalu) [inline docs/src/proposals/leader-qos.md:54]: not directly related to this change, but why continue to use PacketBatch as opposed to splitting it apart into a deque + trying to build thicc batches of 128 by traversing all pending packets until the end is reached or 128 are locked?

something to do with memory management wrt shifting a bunch of elements?

Comment (taozhu-chicago) [reply docs/src/proposals/leader-qos.md:54]: There prolly have few advantages of using PacketBatch, one of them is to support sigverify at gpu. @sakridge can shed more light on it.

@carllin is exploring the very idea (breaking packet_batch into packets within banking_stage) atm.

Comment (buffalu) [reply docs/src/proposals/leader-qos.md:54]: makes sense at sigverify, talking more about after sigverify in banking stage

Comment (carllin) [reply docs/src/proposals/leader-qos.md:54]: @buffalu I have a draft with these changes here: https://github.com/solana-labs/solana/pull/24187 if you want to take a look

Comment (buffalu) [reply docs/src/proposals/leader-qos.md:54]: oh lets fucking go!

Comment (carllin) [reply docs/src/proposals/leader-qos.md:54]: Yeah I've run it through some of our automated perf test and the local banking benches, and performance seems in line with master. 

Feel free to suggest/run some other tests you deem interesting

**Pull request closed without merging**